### ISSUE TITLE
[go_router] !keyReservation.contains(key) assert fails when trying to push a different instance of a similar route over a current one in the stack.

### ISSUE THREAD

**Issue by KimAloy** [flutter/flutter#107010]
I'm using go_router's  context.push method in all routes, for example context.push('/post').
In my navigation drawer I have 3 routes namely "Home", "My Account", "Post Advert". The initial route is "Home". When I tap "Post Advert", it navigates to the "Post Advert" fine. Then I tap "My Account", it too navigates to "My Account" fine. But when I tap on "Post Advert" page again,  it throws the error "'!keyReservation.contains(key)': is not true". 

Looks like another navigation package had the same issue. Please note that I'm submitting this issue about go_router. I'm pasting the link for reference here to perhaps aid in figuring out the solution. Here's the link https://github.com/Milad-Akarie/auto_route_library/issues/492.

Again, this is an issue I'm facing for go_router navigation package.

**Comment by motasimfuad** [comment]
Did you able to solve the issue?

I am facing the same issue. however, I used context.pushNamed.
I understand it's about the duplicate global key, but not sure how to fix it.

**Comment by KimAloy** [comment]
No, I'm waiting for the flutter team to solve it. Please let's keep in
touch & update me if you solve it. I'll be updating you also in case I find
a work around.

**Comment by darshankawar** [comment]
@KimAloy 
Can you provide `flutter doctor -v` and [complete minimal reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example) that shows the behavior you are reporting.
Thanks.

**Comment by mamuy-gu** [comment]
In version < 4.0 (eg 3.1.1), when a route is pushed, the number of pushes for fullpath is calculated and the pageKey parameter is determined from the number of those pushes.
In version > 4.0, after the GoRouterDelegate refactoring, this logic has been omitted, which now does not allow pushing one page onto the stack more than once, since the value of pageKey is always equal to fullpath
![image](https://user-images.githubusercontent.com/59157948/177131508-050f5eb3-5fd5-414d-80bf-ec90d626aff8.png)

**Comment by FusoraTech** [comment]
Related: https://github.com/flutter/flutter/issues/107045

**Comment by KimAloy** [comment]
@darshankawar 
I've tried to create an MRE using the same list of routers offline, but it seems to works fine. Is it because there's an issue when using go_router with Firestore. Perhaps @motasimfuad could reproduce the bug. I've failed for the past 2 days, but when I run the project code and try to navigate as I wish using the navigation drawer stacking same route. It errors "Assertion failed: file:///C:flutter/packages/flutter/lib/src/widgets/navigator.dart:3470:18 !keyReservation.contains(key) is not true"

Here's the flutter doctor -v if it helps

[√] Flutter (Channel stable, 3.0.0, on Microsoft Windows [Version 10.0.19044.1586], locale en-UG)
    • Flutter version 3.0.0 at C:\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ee4e09cce0 (8 weeks ago), 2022-05-09 16:45:18 -0700
    • Engine revision d1b9a6938a
    • Dart version 2.17.0
    • DevTools version 2.12.2

Checking Android licenses is taking an unexpectedly long time...[√] Android toolchain - develop for Android devices (And
roid SDK version 32.1.0-rc1)
    • Android SDK at C:\Users\MadCoder\AppData\Local\Android\sdk
    • Platform android-32, build-tools 32.1.0-rc1
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[X] Visual Studio - develop for Windows
    X Visual Studio not installed; this is necessary for Windows development.
      Download at https://visualstudio.microsoft.com/downloads/.
      Please install the "Desktop development with C++" workload, including all of its default components

[√] Android Studio (version 2021.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)

[√] VS Code (version 1.67.0)
    • VS Code at C:\Users\MadCoder\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.40.0

[√] Connected device (4 available)
    • SM G9880 (mobile) • emulator-5554 • android-x64    • Android 7.1.2 (API 25)
    • Windows (desktop) • windows       • windows-x64    • Microsoft Windows [Version 10.0.19044.1586]
    • Chrome (web)      • chrome        • web-javascript • Google Chrome 103.0.5060.66
    • Edge (web)        • edge          • web-javascript • Microsoft Edge 103.0.1264.44
    ! Device 127.0.0.1:5555 is offline.

[√] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
